bpf/cpu/pyperf.bpf.c

```diff
@@ -0,0 +1,439 @@
+  LOG("interpreter_info->thread_state_addr 0x%llx", interpreter_info->thread_state_addr);
+  int err = bpf_probe_read_user(&state->thread_state, sizeof(state->thread_state), (void *)(long)interpreter_info->thread_state_addr);
+  if (err != 0) {
+    LOG("[error] bpf_probe_read_user failed with %d", err);
+    goto submit_event;
+  }
+  if (state->thread_state == 0) {
+    LOG("[error] thread_state was NULL");
+    goto submit_event;
+  }
+  LOG("thread_state 0x%llx", state->thread_state);
+
+  struct task_struct *task = (struct task_struct *)bpf_get_current_task();
+  long unsigned int tls_base = read_tls_base(task);
+  LOG("tls_base 0x%llx", (void *)tls_base);
+
+  GET_OFFSETS();
+
+  // Fetch the thread id.
+
+  // Python 3.11+ uses native_thread_id.
+  if (offsets->py_thread_state.native_thread_id > 0) {
+    u64 thread_id = 0;
+    LOG("offsets->py_thread_state.native_thread_id %d", offsets->py_thread_state.native_thread_id);
+    bpf_probe_read_user(&thread_id, sizeof(thread_id), state->thread_state + offsets->py_thread_state.native_thread_id);
+    LOG("thread_id %d", thread_id);
+    if (thread_id != tid) {
+      LOG("[error] thread_id %d != tid %d", thread_id, tid);
+      goto submit_event;
+    }
+  } else {
+    LOG("offsets->py_thread_state.thread_id %d", offsets->py_thread_state.thread_id);
+    pthread_t pthread_id;
+    bpf_probe_read_user(&pthread_id, sizeof(pthread_id), state->thread_state + offsets->py_thread_state.thread_id);
+    LOG("pthread_id %lu", pthread_id);
+    // 0x10 = offsetof(tcbhead_t, self) for glibc.
+    pthread_t current_pthread_id;
+    bpf_probe_read_user(&current_pthread_id, sizeof(current_pthread_id), (void *)(tls_base + 0x10));
+    LOG("current_pthread_id %lu", current_pthread_id);
+    if (pthread_id != current_pthread_id) {
+      LOG("[error] pthread_id %lu != current_pthread_id %lu", pthread_id, current_pthread_id);
+      goto submit_event;
+    }
+    state->current_pthread = current_pthread_id;
+  }
+
+  // Get pointer to top frame from PyThreadState.
+
+  if (offsets->py_thread_state.frame > -1) {
+    LOG("offsets->py_thread_state.frame %d", offsets->py_thread_state.frame);
+    bpf_probe_read_user(&state->frame_ptr, sizeof(void *), state->thread_state + offsets->py_thread_state.frame);
+  } else {
+    LOG("offsets->py_thread_state.cframe %d", offsets->py_thread_state.cframe);
+    void *cframe;
+    bpf_probe_read_user(&cframe, sizeof(cframe), (void *)(state->thread_state + offsets->py_thread_state.cframe));
+    if (cframe == 0) {
+      LOG("[error] cframe was NULL");
+      goto submit_event;
+    }
+    LOG("cframe 0x%llx", cframe);
+
+    LOG("offsets->py_cframe.current_frame %d", offsets->py_cframe.current_frame);
+    bpf_probe_read_user(&state->frame_ptr, sizeof(state->frame_ptr), (void *)(cframe + offsets->py_cframe.current_frame));
+  }
+  if (state->frame_ptr == 0) {
+    LOG("[error] frame_ptr was NULL");
+    goto submit_event;
```

Comment:
Perhaps worth a comment if this condition can only be hit when we are done with the Python stack (I assume it's the NULL pointer at the end of the linked list of frames?)

Reply:
💯 I will add more explanation in the comments.

Reply:
> (I assume it's the NULL pointer at the end of the linked list of frames?)

Noop, this is something else.